sockshop/catalogue/service_test.go

```diff
@@ -111,49 +185,44 @@ func TestFixedServiceGet(t *testing.T) {
 }
 
 func TestFixedServiceTags(t *testing.T) {
-	// This test is kind of pointless, I guess.
-	s := NewFixedService(socks, tags)
-	if want, have := tags, s.Tags(); !reflect.DeepEqual(want, have) {
-		t.Errorf("Tags(): want %v, have %v", want, have)
+	db, mock, err := sqlmock.New()
+	if err != nil {
+		t.Fatalf("an error '%s' was not expected when opening stub database connection", err)
 	}
-}
+	defer db.Close()
 
-func TestFilter(t *testing.T) {
-	for _, testcase := range []struct {
-		tags []string
-		want []Sock
-	}{
-		{[]string{}, []Sock{s1, s2, s3, s4, s5}}, // no tags, return all
-		{[]string{""}, []Sock{}},
-		{[]string{"true"}, []Sock{s1, s2, s3, s4, s5}}, // special case
-		{[]string{"odd"}, []Sock{s1, s3, s5}},
-		{[]string{"even"}, []Sock{s2, s4}},
-		{[]string{"odd", "even"}, []Sock{}}, // impossibility result
-		{[]string{"prime", "odd"}, []Sock{s1, s3, s5}},
-		{[]string{"prime", "even"}, []Sock{s2}}, // AND semantics
-	} {
-		if want, have := testcase.want, filter(socks, testcase.tags); !reflect.DeepEqual(want, have) {
-			t.Errorf("filter(%v): want %s, have %s", testcase.tags, printIDs(want), printIDs(have))
-		}
-	}
-}
+	var cols []string = []string{"name"}
 
-func TestSortBy(t *testing.T) {
-	for _, testcase := range []struct {
-		input []Sock
-		order string
-		want  []Sock
-	}{
-		{[]Sock{s5, s4, s3, s2, s1}, "undefined, should be no-op", []Sock{s5, s4, s3, s2, s1}},
-		{[]Sock{s5, s4, s3, s2, s1}, "id", []Sock{s1, s2, s3, s4, s5}},
-		{[]Sock{s1, s2, s3, s4, s5}, "tag", []Sock{s4, s1, s2, s3, s5}},
-	} {
-		if want, have := testcase.want, sortBy(testcase.input, testcase.order); !reflect.DeepEqual(want, have) {
-			t.Errorf("sortBy(%s): want %s, have %s", testcase.order, printIDs(want), printIDs(have))
-		}
+	mock.ExpectQuery("SELECT name FROM Tag").WillReturnRows(sqlmock.NewRows(cols).
+		AddRow(tags[0]).
+		AddRow(tags[1]).
+		AddRow(tags[2]))
+
+	s := NewFixedService(db)
+
+	have, err := s.Tags()
+	if err != nil {
+		t.Errorf("Tags(): %v", err)
+	}
+	if !reflect.DeepEqual(tags, have) {
+		t.Errorf("Tags(): want %v, have %v", tags, have)
 	}
 }
 
+// TODO How to mock this?
```

Comment:
You're having trouble because your FixedService takes a concrete *sql.DB. A better way to do this, I think, is to have it take an intermediate `type Database interface` with methods for each of the queries that you'll perform. Then, you can wrap a *sql.DB with some very straightforward code which implements that interface, and program the service against the interface rather than the struct, i.e. against a behavioral rather than physical contract. Finally, you can test the intermediate wrapper using simple SQL mocks, similar to what you've done here; and you can test the service using a mock implementation of the interface with fixed data.